File: typescript/kata-machine/src/day1/Stack.ts
type QNode<T> = {
    value: T;
    prev?: QNode<T>;
};

export default class Stack<T> {
    public length: number;
    private head?: QNode<T>;

    constructor() {
        this.length = 0;
        this.head = undefined;
    }

    push(item: T): void {
        const node = { value: item } as QNode<T>;

        this.length++;

        if (!this.head) {
            this.head = node;
            return;
        }

        node.prev = this.head;
        this.head = node;
    }

    pop(): T | undefined {
        this.length = Math.max(0, this.length - 1);

        if (this.length === 0) {
            const head = this.head;
            this.head = undefined;
            return head?.value;
        }

        const head = this.head as QNode<T>;
        this.head = head.prev;

        return head.value;
    }

    peek(): T | undefined {
        return this.head?.value;
    }
}

File: cpp/algo/src/structures/Stack.cpp
#include "iostream"
#include "Stack.h"
#include "../main.h"

using namespace std;

void stackExample() {
    cout << "Stack" << endl;

    Stack<int> stack;

    cout << "Tamanho da stack: " << stack.size() << endl;

    stack.push(1);
    stack.push(2);
    stack.push(3);
    stack.push(4);
    stack.push(5);
    stack.push(6);
    stack.push(7);
    stack.push(8);
    stack.push(9);
    stack.push(10);

    cout << stack.peek() << endl;

    while (stack.size() > 0) {
        cout << stack.pop() << endl;
    }

    cout << endl;
}
